File: flutter/lib/core/services/user_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/user.dart';

/// Service utilisateur — miroir de UserService Angular.
class UserService {
  UserService({FirebaseFirestore? firestore})
      : _db = firestore ?? FirebaseFirestore.instance;

  final FirebaseFirestore _db;

  /// Élèves rattachés à un bénévole (benevoleId == benevoleId)
  Stream<List<AppUser>> getElevesOfBenevole(String benevoleId) {
    return _db
        .collection('users')
        .where('benevoleId', isEqualTo: benevoleId)
        .snapshots()
        .map((snap) => snap.docs
            .map(AppUser.fromDoc)
            .toList());
  }

  /// Retourne un utilisateur par son UID en temps réel.
  Stream<AppUser?> getUserById(String uid) {
    return _db
        .collection('users')
        .doc(uid)
        .snapshots()
        .map((snap) => snap.exists ? AppUser.fromDoc(snap) : null);
  }

  /// Met à jour le profil (nom, prénom) dans Firestore.
  Future<void> updateProfile(String uid,
      {required String nom, required String prenom}) async {
    await _db.collection('users').doc(uid).update({
      'nom': nom,
      'prenom': prenom,
    });
  }

  /// Met à jour l'URL de la photo de profil dans Firestore.
  Future<void> updatePhotoUrl(String uid, String photoUrl) async {
    await _db
        .collection('users')
        .doc(uid)
        .update({'photoUrl': photoUrl});
  }
}

File: angular/src/app/core/services/user.service.ts
import { inject, Injectable } from '@angular/core';
import {
  Firestore,
  collection,
  collectionData,
  doc,
  getDoc,
  updateDoc,
  query,
  where,
} from '@angular/fire/firestore';
import { from, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { User } from '../models/user.model';

@Injectable({ providedIn: 'root' })
export class UserService {
  private firestore = inject(Firestore);

  /**
   * Retourne les élèves rattachés à un bénévole.
   */
  getElevesOfBenevole$(benevoleId: string): Observable<User[]> {
    const q = query(
      collection(this.firestore, 'users'),
      where('benevoleId', '==', benevoleId)
    );
    return collectionData(q, { idField: 'id' }) as Observable<User[]>;
  }

  /**
   * Retourne un utilisateur par son UID.
   */
  getUserById$(uid: string): Observable<User | null> {
    return from(getDoc(doc(this.firestore, `users/${uid}`))).pipe(
      map(snap => snap.exists() ? ({ id: snap.id, ...snap.data() } as User) : null)
    );
  }

  /**
   * Met à jour la photo de profil d'un utilisateur.
   * @param uid UID Firestore de l'utilisateur
   * @param photoUrl Data-URL base64 de l'image compressée
   */
  updatePhotoUrl(uid: string, photoUrl: string): Promise<void> {
    return updateDoc(doc(this.firestore, `users/${uid}`), { photoUrl });
  }

  /**
   * Met à jour le prénom et le nom d'un utilisateur.
   */
  updateProfile(uid: string, data: { prenom: string; nom: string }): Promise<void> {
    return updateDoc(doc(this.firestore, `users/${uid}`), { ...data });
  }
}
